### DISCUSSION TITLE
Support exposing database port to docker host

### DISCUSSION THREAD

**Discussion by gclawes** [nextcloud/all-in-one#1066] [Ideas] [answered]
It would be useful to expose the database port to the docker host in the same way `APACHE_IP_BINDING`/`APACHE_PORT` work.  This would allow external backup solutions like [borgmatic](https://torsion.org/borgmatic) to perform [database dumps](https://torsion.org/borgmatic/docs/how-to/backup-your-databases/).

It is possible to [communicate with the database over the docker network](https://github.com/nextcloud/all-in-one/#phpmyadmin-adminer-or-pgadmin), though there are cases where an external backup tool like borgmatic does not necessarily run in docker, so accessing the database port over locahost would be useful.

**Comment by szaimen** [comment, marked as answer]
Hm... I would always recommend to stop the containers of AIO correctly before doing any backup externally. This has the advantage that AIO will also create a database dump automatically upon container shut down and you do not need to do it manually. See https://github.com/nextcloud/all-in-one#how-to-stopstartupdate-containers-or-trigger-the-daily-backup-from-a-script-externally on how to stop the containers and start them again.

I guess you should be able to use [this feature](https://torsion.org/borgmatic/docs/how-to/add-preparation-and-cleanup-steps-to-backups/) of borgmatic to run a shell script or shell command before and after the backup.

**Reply by gclawes** [reply]
Yep, I'm using that borgmatic feature to place nextcloud into maintenance mode, but I'll look into stopping the containers entirely.  Thanks!

**Reply by Zoey2936** [reply]
Just a small question, why are you using borgmatic? What's better than using the inbuilt Borg function?

**Reply by gclawes** [reply]
Though stopping with `DAILY_BACKUP=1` will cause AIO to use it's internal borg backup mechanism, correct?   Is that required to do the database dump on container shutdown, or will that happen even withouth `DAILY_BACKUP=1`?

**Reply by gclawes** [reply]
@Zoey2936  borgmatic can write to remote borg repositories.  As far as I can tell there's no way to do that with the built-in mechanism.

**Reply by Zoey2936** [reply]
> Though stopping with `DAILY_BACKUP=1` will cause AIO to use it's internal borg backup mechanism, correct? Is that required to do the database dump on container shutdown, or will that happen even withouth `DAILY_BACKUP=1`?

If you just want the to stop the containers and export the database dump, it should be enough to run it with `STOP_CONTAINERS=1`
But have a note on #1067

**Reply by Zoey2936** [reply]
> @Zoey2936 borgmatic can write to remote borg repositories. As far as I can tell there's no way to do that with the built-in mechanism.

If I know correctly, BorgBackup has also an integration for this (remote borg repos), but this function is **not** implemented in aio
https://borgbackup.readthedocs.io/en/stable/quickstart.html#remote-repositories

**Reply by gclawes** [reply]
> > @Zoey2936 borgmatic can write to remote borg repositories. As far as I can tell there's no way to do that with the built-in mechanism.
> 
> If I know correctly, BorgBackup has also an integration for this (remote borg repos), but this function is **not** implemented in aio https://borgbackup.readthedocs.io/en/stable/quickstart.html#remote-repositories

Right.  Borgmatic is just a wrapper around borgbackup, it's effectively a way to have a nice borg config file.
